lib_eio/stream.ml

```diff
@@ -94,6 +94,45 @@ module Locking = struct
       Mutex.unlock t.mutex;
       Some v
 
+  let select_of_many streams_fns =
+    let finished = Atomic.make false in
+    let cancel_fns = ref [] in
+    let add_cancel_fn fn = cancel_fns := fn :: !cancel_fns in
+    let cancel_all () = List.iter (fun fn -> fn ()) !cancel_fns in
+    let wait ctx enqueue (t, f) = begin
+      Mutex.lock t.mutex;
+      (* First check if any items are already available and return early if there are. *)
+      if not (Queue.is_empty t.items)
+      then (
+        cancel_all ();
+        Mutex.unlock t.mutex;
+        enqueue (Ok (f (Queue.take t.items))))
```

Comment:
Another domain could still have set `finished` to true using a different stream by the time this runs.

(also, you can't use `Queue.take` after unlocking)

Reply:
Apologies for missing this, it should have been obvious! Please take another look, it seemed not too difficult to fix.